### DISCUSSION TITLE
TDD vs BDD vs ATDD

### DISCUSSION THREAD

**Discussion by hochan222** [transcendence42/javascript-archive#19] [Topic]
TDD vs BDD vs ATDD

**Comment by yechoi42** [comment]
## TDD(test-driven development) - 테스트 주도 개발
테스트 주도 개발(test-driven development, TDD)은 매우 짧은 개발 사이클을 반복하는 소프트웨어 개발 프로세스 중 하나이다. 개발자는 우선 자동화된 테스트 케이스를 작성하고, 그 케이스를 통과하기 위한 최소한의 코드를 생성, 리팩토링 하는 개발 사이클을 가진다. 테스트를 최우선으로 "작동하는 깔끔한 코드(clean code that works)"를 만드는 것이 테스트 주도 개발의 궁극적인 목표이다. 

### TDD 사이클
- red : 실패하는 작은 테스트를 작성한다.
- green : 테스트를 통과하는 코드를 작성한다. (퀄리티보다는 일단 통과하는 것을 목표로 작성한다.)
- refactoring : 테스트를 통과한 코드를 리팩토링 한다. (퀄리티가 떨어지는 코드를 이 단계에서 리팩토링 한다.)

**Comment by hochan222** [comment]
## TDD (Test-driven development)

- [이규원이 말하는 현실 세상의 TDD : 안정감을 주는 코드 작성 방법.](https://www.youtube.com/watch?v=gX9PkEibWsM)

**"TDD는 짧은 주기로 피드백을 받아가며 프로그래머가 코드를 늘려가는 개발 방법론"**

엔지니어링은 결점을 없애는 것이 아니라 결점 혹은 한 문제를 다른 문제로 전환하는 일이다.
엔지니어링에서 지금 상황에 우리에게 피해를 많이 주는 문제를 피해가 조금 덜 받는 문제로 치환해주는 일이 대부분이다. 그래서 무결한 설계보다는 지금 상황에 유리한 설계를 하는 자세를 가져야한다. TDD도 마찬가지로 TDD를 통해서 무결한 시스템을 만들자는 접근보다는 겪고 있는 부분을 `어떻게 하면 다른 부분과 trade-off 할 수 있을까?` 그래서 우리가 자원을 아끼고, 더 빨리 개발하고, 더 중요한 가치에 집중 할 수 있을까에 집중해야한다.

<img src="https://user-images.githubusercontent.com/22424891/120251806-a6530c00-c2bd-11eb-8954-0d59344cf440.png" height="300px" />

Test-driven development 로 테스트 코드를 먼저 만들고 테스트 코드를 통과시키는 코드를 작성하는 소프트웨어 개발 프로세스.

## BDD (Behaviour-Driven Development)

BDD는 (Behavior Driven Development )로 TDD를 근간으로 파생된 개발 방법으로, TDD에서 한 발 더 나아가 테스트케이스 자체가 요구사양이 되도록 하는 개발방법이다. (Behavior 기반으로 비지니스 요구 사항에 집중한 테스트 코드를 작성)

BDD를 통해 개발을 하게 된다면 테스트 메소드의 이름을  **"이  클래스가 어떤 행위를 해야한다 (should do someting)"** 라는 식의 문장으로 작성하여 행위에 대한 테스트에 집중할 수  있습니다.

TDD는 코드 자체에 집중한 테스트 코드를 작성하는 반면, BDD는 시나리오를 기반으로 테스트 케이스를 작성하며 함수 단위 테스트를 권장하지 않는다. 이 시나리오는 개발자가 아닌 사람이 봐도 이해할 수 있을 정도의 레벨을 권장한다. 하나의 시나리오는 Given, When, Then 구조를 가지는 것을 기본 패턴으로 권장하며 각 절의 의미는 다음과 같다.

- Feature : 테스트에 대상의 기능/책임을 명시한다.
- Scenario : 테스트 목적에 대한 상황을 설명한다.
- Given : 시나리오 진행에 필요한 값을 설정한다.
- When : 시나리오를 진행하는데 필요한 조건을 명시한다.
- Then : 시나리오를 완료했을 때 보장해야 하는 결과를 명시한다.

테스트 대상의 상태 변화를 테스트하는 것이다.

## ATDD (Acceptance Test Driven Development)

<img src="https://user-images.githubusercontent.com/22424891/120254097-fe8d0c80-c2c3-11eb-89b3-98ccbfc1c7a9.png" height="300px"/>

ATDD는 인수 테스트를 먼저 작성한 다음 기능 개발을 하는 개발방법론이다. 인수 테스트는 고객이 만들어진 소프트웨어를 실제로 사용하기 전에 테스트 하는 것을 말한다. 4가지 ACTION(Discuss, Distill, Develop, Demo)과 4가지 산출물(User Story, Accpetance Cretieria, Tests, Working software), 하나의 결과(Business Value)로 표현되고 있다.

범위가 너무 방대하기 떄문에 자세한건 https://velog.io/@windtrip/ATDDAcceptance-Test-Driven-Development 참고.
